### PR TITLE
Refactor child query

### PR DESCRIPTION
**Pull request by maeb**

### CODE REVIEW

File: pkg/server/warcserver/daterange.go [hunk 1]
```diff
@@ -18,25 +18,24 @@ package warcserver
...
 	to   string
 }
 
-func parseDateRange(from, to string) *dateRange {
-	from = fmt.Sprintf("%s%0*d", from, 14-len(from), 0)
-	to = fmt.Sprintf("%s%.*s", to, 14-len(to), "99999999999999")
-	d := &dateRange{from: from, to: to}
-	return d
+// contains returns true if the timestamp ts contained by the bounds defined by the DateRange d.
+func (d DateRange) contains(ts string) bool {
```
Comment (Avokadoen) [inline pkg/server/warcserver/daterange.go:29]: This does not seem to work as intended: https://play.golang.org/p/ITLMRqt30ZU
It's also a good unit test contender

File: proto/Makefile [hunk 2]
```diff
@@ -0,0 +1,24 @@
+PROTOC_VERSION=3.11.4
```
Comment (Avokadoen) [inline proto/Makefile:1]: Why do we pull inn this dependency?

File: pkg/server/handlers/childhandler.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	return ChildHandler(children, timeout, responseHandler)
+}
+
+func FirstHandler(children []*url.URL, timeout time.Duration) http.Handler {
+	responseHandler := func(w http.ResponseWriter, r *http.Request, responses <-chan *http.Response) {
+		written := false
+		for response := range responses {
+			resp := response
+			if resp.StatusCode < 400 {
```
[the remark this answers is not in the document]
Comment (Avokadoen) [reply pkg/server/handlers/childhandler.go:41]: And maybe assert body size to be sure there is a content response

**Comment by Avokadoen** [comment]
Got error on serve:
```
$ warcserve serve
....
....
INFO[0000] indexing testdata/example.warc               
INFO[0000] Finished indexing warcdb/cdx-index/DISCARD. Found: 0 records in: 19.356668ms 
ERRO[0000] Key not found                                
ERRO[0000] Error: gzip: invalid header, rec num: 0, Offset 468013 
INFO[0000] Finished indexing warcserver. Found: 0 records in: 22.993216ms 
INFO[0000] indexing testdata/IAH-20080430204825-00000-blackbook.warc 
FATA[0000] Unreachable   
```

File: pkg/server/handlers/childhandler.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+				if err != nil {
+					log.WithField("uri", resp.Request.RequestURI).WithField("status", resp.Status).Errorln(err)
+				} else {
+					written = true
+				}
+			}
+			_ = resp.Body.Close()
+		}
+		if !written {
+			http.Error(w, "", http.StatusOK)
```
[the remark this answers is not in the document]
Comment (maeb) [reply pkg/server/handlers/childhandler.go:29]: I don't know, will have to dig into pywb's warcserver to know what it returns when there are no results matching a search.

File: pkg/server/handlers/childhandler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maeb) [reply pkg/server/handlers/childhandler.go:41]: I guess we should make an issue about things like this. If the original webserver returned a status code I guess pywb should respond with the same one even tough it is >= 400 even.

File: pkg/server/handlers/childhandler.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+			close(responses)
+		}()
+
+		for _, childUrl := range children {
+			req := r.Clone(ctx)
+			req.RequestURI = ""
+			req.URL = BuildChildURLString(childUrl, req.URL)
+			go func() {
+				defer wg.Done()
+				//nolint:bodyclose
```
[the remark this answers is not in the document]
Comment (maeb) [reply pkg/server/handlers/childhandler.go:89]: Will do.
```suggestion
				// The response consumer is responsible for closing the response body
				//nolint:bodyclose
```

File: proto/Makefile (hunk 2 above)
Comment (maeb) [reply proto/Makefile:1]: I don't see this as a dependency per se. We must ensure we git ignore any artifacts that are created in the build process.

File: pkg/server/warcserver/daterange.go (hunk 1 above)
Comment (maeb) [reply pkg/server/warcserver/daterange.go:29]: I agree this should work in isolation. In the context we are using it the members of the `DateRange` struct is always padded with 0's or 9's to a length of 14.

Comment (Avokadoen) [reply pkg/server/warcserver/daterange.go:29]: An issue should be created when this gets merged about making the function more "fool proof"

File: proto/Makefile (hunk 2 above)
Comment (Avokadoen) [reply proto/Makefile:1]: Conclusion is that proto is a needed dependency currently

**Review by Avokadoen — APPROVED** [review]

**Pull request merged**